src/main/java/spoon/support/visitor/java/JavaReflectionVisitorImpl.java

```diff
@@ -308,14 +309,18 @@ public <T> void visitConstructor(Constructor<T> constructor) {
 		for (Annotation annotation : constructor.getDeclaredAnnotations()) {
 			visitAnnotation(annotation);
 		}
-		int nrEnclosingClasses = getNumberOfEnclosingClasses(constructor.getDeclaringClass());
-		for (RtParameter parameter : RtParameter.parametersOf(constructor)) {
-			//ignore implicit parameters of enclosing classes
-			if (nrEnclosingClasses > 0) {
-				nrEnclosingClasses--;
+		RtParameter[] parametersOf = RtParameter.parametersOf(constructor);
+		Parameter[] parameters = constructor.getParameters();
+		for (int i = 0; i < parametersOf.length; i++) {
+			if (parameters[i].isImplicit()) {
```

Comment:
*It's complicated*. Since Java 8, the JLS says that
> 12. A construct emitted by a Java compiler must be marked as *mandated* if it corresponds to a formal parameter declared implicitly in source code

([JLS 13.1](https://docs.oracle.com/javase/specs/jls/se8/html/jls-13.html#jls-13.1))

However, javac only does this when compiling with the `-parameters` flag. See https://mail.openjdk.java.net/pipermail/compiler-dev/2022-May/019783.html for more context.

So ih theory, it's the best way to figure out if a parameter is implicit, but in practice, it would require the compiler to keep to the spec.

Reply:
I see. Then I suggest extracting the implicitness check into a separate method and put all of the relevant information required to understand why it's so funky in the Javadoc (for example the stuff you wrote above). Something like this:

```java
/**
 * Determine if the parameter at parameterPosition is is implicit.
 *
 * This is a great description of why this is so weird.
 */
private boolean isImplicitParameter(Parameter parameter, Constructor constructor, boolean isFirstParameter) {
    // do isImplicit() and "best guess" checks
}
```

It doesn't have to be exactly like that, but just provide any information required into the method such that it can perform the checks. The implementation of `Constructor.getParameters()` looks rather heavy so we probably don't want to call it multiple times.

There are a lot of places in Spoon where weird corner cases aren't adequately explained, so I'm trying to make a push for... explaining them.

Reply:
I extracted the relevant code into its own method now.